sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/MapAggregate.scala

```diff
@@ -0,0 +1,310 @@
+      // return empty map
+      ArrayBasedMapData(Map.empty)
+    } else {
+      // sort the result to make it more readable
+      val sorted = buffer match {
+        case stringDigest: StringMapDigest => TreeMap[UTF8String, Long](stringDigest.bins.toSeq: _*)
+        case numericDigest: NumericMapDigest => TreeMap[Double, Long](numericDigest.bins.toSeq: _*)
+      }
+      if (sorted.isEmpty) {
+        // don't have non-null values
+        null
+      } else {
+        ArrayBasedMapData(sorted.keys.toArray, sorted.values.toArray)
+      }
+    }
+  }
+
+  override def serialize(buffer: MapDigest): Array[Byte] = {
+    buffer.serialize()
+  }
+
+  override def deserialize(bytes: Array[Byte]): MapDigest = {
+    MapDigest.deserialize(child.dataType, bytes)
+  }
+
+  override def createAggregationBuffer(): MapDigest = {
+    child.dataType match {
+      case StringType => StringMapDigest()
+      case _ => NumericMapDigest()
+    }
+  }
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): MapAggregate = {
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+  }
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): MapAggregate = {
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+  }
+
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = {
+    child.dataType match {
+      case StringType => MapType(StringType, LongType)
+      case _ => MapType(DoubleType, LongType)
+    }
+  }
+
+  override def children: Seq[Expression] = Seq(child, numBinsExpression)
+
+  override def prettyName: String = "map_aggregate"
+}
+
+trait MapDigest {
+  // Mark this MapDigest invalid when the size of the hashmap (ndv of the column) exceeds numBins
+  var isInvalid: Boolean
+  def update(dataType: DataType, value: Any, numBins: Int): Unit
+  def merge(otherDigest: MapDigest, numBins: Int): Unit
+  def clear(): Unit
+  def serialize(): Array[Byte]
+}
+
+abstract class MapDigestBase[T] extends MapDigest {
+  val bins: mutable.HashMap[T, Long]
+
+  // Update bins and clear it when its size exceeds numBins.
+  def updateMap(value: T, numBins: Int): Unit = {
+    mergeMap(mutable.HashMap(value -> 1L), numBins)
+  }
+
+  // Merge two maps and clear bins when its size exceeds numBins.
+  override def merge(otherDigest: MapDigest, numBins: Int): Unit = {
+    val otherMap = otherDigest.asInstanceOf[MapDigestBase[T]].bins
+    mergeMap(otherMap, numBins)
+  }
+
+  def mergeMap(otherMap: mutable.HashMap[T, Long], numBins: Int): Unit = {
+    otherMap.foreach { case (key, value) =>
+      if (bins.contains(key)) {
+        bins.update(key, bins(key) + value)
+      } else {
+        if (bins.size >= numBins) {
+          isInvalid = true
+          bins.clear()
+          return
+        } else {
+          bins.put(key, value)
+        }
+      }
+    }
+  }
+
+  override def clear(): Unit = bins.clear()
+
+  override def serialize(): Array[Byte] = {
```

Comment:
should we just serialize using Kryo rather than writing our own serializer here?